core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java

```diff
@@ -82,41 +100,90 @@
+        EntitySpec<? extends Application> spec = createEntitySpec(mgmt, yaml);
+        return createUnstarted(mgmt, spec);
+    }
+
+    public static EntitySpec<? extends Application> createEntitySpec(ManagementContext mgmt, String yaml) {
+        return createEntitySpec(mgmt, new StringReader(yaml));
+    }
+
+    public static EntitySpec<? extends Application> createEntitySpec(ManagementContext mgmt, Reader yaml) {
+        BrooklynClassLoadingContext loader = createDefaultLoader(mgmt);
+        return createEntitySpec(mgmt, yaml, loader);
+    }
+
     @SuppressWarnings("unchecked")
-    public static <T extends Application> T createUnstarted(ManagementContext mgmt, AssemblyTemplate at) {
-        CampPlatform camp = getCampPlatform(mgmt).get();
-        AssemblyTemplateInstantiator instantiator;
+    public static EntitySpec<? extends Application> createEntitySpec(ManagementContext mgmt, Reader yaml, BrooklynClassLoadingContext loader) {
+        return (EntitySpec<? extends Application>) createSpec(mgmt, yaml, loader);
+    }
+
+    private static EntitySpec<? extends Application> createEntitySpec(String symbolicName, CampPlatform camp,
+                                                                      DeploymentPlan plan, BrooklynClassLoadingContext loader) {
+        AssemblyTemplate at = registerDeploymentPlan(camp, plan, loader);
+        AssemblyTemplateInstantiator instantiator = newInstantiator(at);
+        if (instantiator instanceof AssemblyTemplateSpecInstantiator) {
+            return (EntitySpec<? extends Application>) ((AssemblyTemplateSpecInstantiator)instantiator).createNestedSpec(at, camp, loader, MutableSet.of(symbolicName));
+        } else {
+            throw new IllegalStateException("Unable to instantiate YAML; incompatible instantiator " + instantiator + " for " + at);
+        }
+    }
+
+    public static AbstractBrooklynObjectSpec<?, ?> createSpec(ManagementContext mgmt, String yaml, BrooklynClassLoadingContext loader) {
+        return createSpec(mgmt, new StringReader(yaml), loader);
+    }
+
+    public static AbstractBrooklynObjectSpec<?, ?> createSpec(ManagementContext mgmt, Reader yaml, BrooklynClassLoadingContext loader) {
+        Preconditions.checkNotNull(mgmt, "mgmt");
+        Preconditions.checkNotNull(mgmt, "yaml");
+        Preconditions.checkNotNull(mgmt, "loader");
+
+        CampPlatform camp = getCampPlatform(mgmt);
+        DeploymentPlan plan = parseDeploymentPlan(camp, yaml);
+        return createSpec(null, mgmt, plan, loader);
+    }
+
+    private static AbstractBrooklynObjectSpec<?, ?> createSpec(String symbolicName, ManagementContext mgmt,
+                                                               DeploymentPlan plan, BrooklynClassLoadingContext loader) {
+        if (isPolicyPlan(plan)) {
+            return createPolicySpec(plan, loader);
+        } else {
+            return createEntitySpec(symbolicName, getCampPlatform(mgmt), plan, loader);
```

Comment:
Here is where things go sideways. The method `createEntitySpec` called here is coming from `BasicBrooklynCatalog`, calling into `createNestedSpec`. You should instead use https://github.com/neykov/incubator-brooklyn/commit/55305ea9bf05598a8ac2c811691008818edbf321#diff-b713fc5de648db4fede03fa71cfd1c6fR315 which calls into `AssemblyTemplateSpecInstantiator.createSpec`.